### ISSUE TITLE
The MNIST Dataset's get_data is broken

### ISSUE THREAD

**Issue by rizar** [bartvm/blocks#100]
It does not have the state parameter, loading happens in `__init__`.

**Comment by bartvm** [comment]
But loading of the data should happen in `__init__`. The MNIST data in memory is not a state i.e. it doesn't change as we read/iterate over the data. More importantly, many different streams can share the same data in memory, and don't need to each have their own copy of the data. That would be very inefficient.

I was unsure about whether the leave the `state` parameter as an argument. There are two ways to deal with the `get_data` interface considering that some `Dataset` implementations won't accept `state` (if they are stateless, like MNIST) or `request` arguments (if they only support simple iteration i.e. one example at a time in order):

* Accept them as keyword arguments with default `None` and raise an error if they are anything else ```python
def get_data(self, state=None, request=None):
    if state is not None:
        raise ValueError("MNIST received a state, but shouldn't!")
    return self.data[request]
```

* Don't accept them as an argument at all, but in that case we need to make sure that the other parts of the library don't pass anything to `get_data` if there is no `state` or `request` to pass

```python
def get_data(self, request):
    return self.data[request]
```

**Comment by rizar** [comment]
Preventing loading multiple times is a good reason, but I think we should keep the invariant that a dataset object is nothing but a light-weight handle to the data and loading happens only when needed. In the current paradigm that would be the `open` method, if we use ideas of DWF that would be the first `get_data`call. I anticipated that we will do like that:

```python
def open(self):
    if not hasattr(self.state):
        self.state = self._load_data()
    return self.state
```

If you want to keep attribute based access by `mnist.X` and `mnist.y`, we could give user an explicit interface to trigger loading, e.g. `mnist.load`. 

As for the `get_data` method, I think it is better to always keep state as a parameter, as it is a part of the interface.

**Comment by bartvm** [comment]
Okay, so I'll change the `get_load` method back to all have the same signature, but just raise an error when the user tries to feed it a `state` that isn't `None`.

I was thinking more along the lines of having a separate method, `load`, in which the dataset needs to perform any action that involves interaction with the outside world e.g. opening files, setting up database connections, etc. This method would support the lazy loading needed for DWF's solution to serialization.

**Comment by bartvm** [comment]
Closed via #92
